Submitted by: <br>
Name: Abraham Chen <br>
ID Number: 12205974 <br>
Course: BS Mathematics with Specialization in Business Applications

# Zomato Exercise

Zomato is a restaurant aggregator website. The platform provides information, menus and user-reviews of restaurants as well as food delivery options from partner restaurants in select cities. You have been provided 3 datasets:

- zomato.csv contains restaurant information like average_cost_for_two, cuisines
- location.csv contains location of the restaurants like address, city, locality & etc
- ratings.csv contains aggregate rating for each restaurant

The different files are connected using the `id` columns which can be found in the different files. The id column is a unique identifier to represent a restaurant. You can use this to **merge** the different files.

In [125]:
import pandas as pd
import numpy as np

1. Merge the zomato, location & ratings file into a signle dataframe and store it to the `merged_df` variable **(3pts)**

Create a DataFrame with zomato.csv:

In [126]:
zomato_df = pd.read_csv("zomato.csv")
zomato_df

id  average_cost_for_two book_url  \
0      6316125                  1000      NaN   
1     18409457                  1300      NaN   
2      6307689                  1600      NaN   
3     18581637                  1000      NaN   
4      6315438                  3000      NaN   
...        ...                   ...      ...   
6825  18635649                  1000      NaN   
6826   6318581                   400      NaN   
6827   6318578                   450      NaN   
6828   6309612                  1000      NaN   
6829   6318582                   600      NaN   

                                        cuisines currency  \
0                                Japanese, Ramen        P   
1                      Korean, Grill, Korean BBQ        P   
2                             American, Filipino        P   
3                                         Korean        P   
4     Seafood, American, Mediterranean, Japanese        P   
...                                          ...      ...   
6825                              Deli, Filipino        P   
6826                                   Fast Food        P   
6827                            Bakery, Filipino        P   
6828                           American, Italian        P   
6829                            Pizza, Fast Food        P   

                          deeplink  \
0      zomato://restaurant/6316125   
1     zomato://restaurant/18409457   
2      zomato://restaurant/6307689   
3     zomato://restaurant/18581637   
4      zomato://restaurant/6315438   
...                            ...   
6825  zomato://restaurant/18635649   
6826   zomato://restaurant/6318581   
6827   zomato://restaurant/6318578   
6828   zomato://restaurant/6309612   
6829   zomato://restaurant/6318582   

                                             events_url  \
0     https://www.zomato.com/manila/mendokoro-ramenb...   
1     https://www.zomato.com/manila/soban-k-town-gri...   
2     https://www.zomato.com/manila/filling-station-...   
3     https://www.zomato.com/manila/romantic-baboy-t...   
4     https://www.zomato.com/manila/niu-by-vikings-b...   
...                                                 ...   
6825  https://www.zomato.com/manila/trader-juans-fil...   
6826  https://www.zomato.com/manila/jollibee-north-c...   
6827  https://www.zomato.com/manila/goldilocks-north...   
6828  https://www.zomato.com/manila/red-buffalo-wing...   
6829  https://www.zomato.com/manila/greenwich-north-...   

                                         featured_image  has_online_delivery  \
0     https://b.zmtcdn.com/data/reviews_photos/971/2...                    0   
1     https://b.zmtcdn.com/data/res_imagery/18290970...                    0   
2     https://b.zmtcdn.com/data/pictures/9/6307689/b...                    0   
3     https://b.zmtcdn.com/data/reviews_photos/cc1/a...                    0   
4     https://b.zmtcdn.com/data/pictures/chains/8/63...                    0   
...                                                 ...                  ...   
6825                                                NaN                    0   
6826  http://b.zmtcdn.com/data/pictures/chains/0/630...                    0   
6827  https://b.zmtcdn.com/data/res_imagery/6300003_...                    0   
6828  https://b.zmtcdn.com/data/res_imagery/6309612_...                    0   
6829  https://b.zmtcdn.com/data/pictures/chains/2/63...                    0   

      has_table_booking  ...  is_delivering_now  \
0                     0  ...                  0   
1                     0  ...                  0   
2                     0  ...                  0   
3                     0  ...                  0   
4                     0  ...                  0   
...                 ...  ...                ...   
6825                  0  ...                  0   
6826                  0  ...                  0   
6827                  0  ...                  0   
6828                  0  ...                  0   
68

Create a DataFrame with ratings.csv:

In [127]:
ratings_df = pd.read_csv("ratings.csv")
ratings_df

id  aggregate_rating rating_color rating_text  votes
0      6316125               4.9       3F7E00   Excellent   1394
1     18409457               4.6       3F7E00   Excellent    602
2      6307689               4.0       5BA829   Very Good    563
3     18581637               4.0       5BA829   Very Good    413
4      6315438               4.6       3F7E00   Excellent    960
...        ...               ...          ...         ...    ...
6825  18635649               2.9       FFBA00     Average      6
6826   6318581               3.0       CDD614     Average      5
6827   6318578               0.0       CBCBC8   Not rated      1
6828   6309612               3.7       9ACD32        Good     96
6829   6318582               0.0       CBCBC8   Not rated      2

[6830 rows x 5 columns]

Create a DataFrame with location.csv:

In [128]:
location_df = pd.read_csv("location.csv")
location_df

id                                            address  \
0      6316125  V. Corporate Center, Soliman Street, Salcedo V...   
1     18409457  Level 3, Greenbelt 3, Ayala Center, Greenbelt,...   
2      6307689  5012 P. Burgos Street, Poblacion, Makati City ...   
3     18581637  CKB Centre, Tomas Morato Avenue Corner Scout R...   
4      6315438  Sixth Floor, SM Aura Premier, C5 Road Corner 2...   
...        ...                                                ...   
6825  18635649   Westgate Center, Filinvest City, Muntinlupa City   
6826   6318581  Zabarte Town Center, Zabarte Corner Camarin Ro...   
6827   6318578  Zabarte Town Center, Zabarte Corner Camarin Ro...   
6828   6309612  Greenhills Town Center, Granada Street, Valenc...   
6829   6318582  Ground Floor, Zabarte Town Center, Zabarte Cor...   

                 city  city_id  country_id   latitude  \
0         Makati City       63         162  14.560313   
1         Makati City       63         162  14.552137   
2         Makati City       63         162  14.563144   
3         Quezon City       63         162  14.633713   
4         Taguig City       63         162  14.546236   
...               ...      ...         ...        ...   
6825  Muntinlupa City       63         162  14.422261   
6826    Caloocan City       63         162  14.756424   
6827    Caloocan City       63         162  14.756406   
6828      Quezon City       63         162  14.610628   
6829    Caloocan City       63         162  14.756382   

                                               locality  \
0                                       Salcedo Village   
1                 Greenbelt 3, San Lorenzo, Makati City   
2                                             Poblacion   
3                                          Tomas Morato   
4     SM Aura Premier, Bonifacio Global City, Taguig...   
...                                                 ...   
6825    Westgate Center, Filinvest City, Mutinlupa City   
6826  Zabarte Town Center, North Caloocan, Caloocan ...   
6827  Zabarte Town Center, North Caloocan, Caloocan ...   
6828     Greenhills Town Center, Valencia, Quezon City    
6829  Zabarte Town Center, North Caloocan, Caloocan ...   

                                       locality_verbose   longitude  zipcode  
0                          Salcedo Village, Makati City  121.024638      NaN  
1     Greenbelt 3, San Lorenzo, Makati City, Makati ...  121.021325      NaN  
2                                Poblacion, Makati City  121.029800   1210.0  
3                             Tomas Morato, Quezon City  121.035026      NaN  
4     SM Aura Premier, Bonifacio Global City, Taguig...  121.054365      NaN  
...                                                 ...         ...      ...  
6825  Westgate Center, Filinvest City, Mutinlupa Cit...  121.033175      NaN  
6826  Zabarte Town Center, North Caloocan, Caloocan ...  121.043329      NaN  
6827  Zabarte Town Center, North Caloocan, Caloocan ...  121.043498      NaN  
6828  Greenhills Town Center, Valencia, Quezon City ...  121.037031      NaN  
6829  Zabarte Town Center, North Caloocan, Caloocan ...  121.043452      NaN  

[6830 rows x 10 columns]

Merge the three DataFrames together:

In [129]:
merged_df = zomato_df.merge(location_df, on="id").merge(ratings_df, on="id")
merged_df

id  average_cost_for_two book_url  \
0      6316125                  1000      NaN   
1     18409457                  1300      NaN   
2      6307689                  1600      NaN   
3     18581637                  1000      NaN   
4      6315438                  3000      NaN   
...        ...                   ...      ...   
6825  18635649                  1000      NaN   
6826   6318581                   400      NaN   
6827   6318578                   450      NaN   
6828   6309612                  1000      NaN   
6829   6318582                   600      NaN   

                                        cuisines currency  \
0                                Japanese, Ramen        P   
1                      Korean, Grill, Korean BBQ        P   
2                             American, Filipino        P   
3                                         Korean        P   
4     Seafood, American, Mediterranean, Japanese        P   
...                                          ...      ...   
6825                              Deli, Filipino        P   
6826                                   Fast Food        P   
6827                            Bakery, Filipino        P   
6828                           American, Italian        P   
6829                            Pizza, Fast Food        P   

                          deeplink  \
0      zomato://restaurant/6316125   
1     zomato://restaurant/18409457   
2      zomato://restaurant/6307689   
3     zomato://restaurant/18581637   
4      zomato://restaurant/6315438   
...                            ...   
6825  zomato://restaurant/18635649   
6826   zomato://restaurant/6318581   
6827   zomato://restaurant/6318578   
6828   zomato://restaurant/6309612   
6829   zomato://restaurant/6318582   

                                             events_url  \
0     https://www.zomato.com/manila/mendokoro-ramenb...   
1     https://www.zomato.com/manila/soban-k-town-gri...   
2     https://www.zomato.com/manila/filling-station-...   
3     https://www.zomato.com/manila/romantic-baboy-t...   
4     https://www.zomato.com/manila/niu-by-vikings-b...   
...                                                 ...   
6825  https://www.zomato.com/manila/trader-juans-fil...   
6826  https://www.zomato.com/manila/jollibee-north-c...   
6827  https://www.zomato.com/manila/goldilocks-north...   
6828  https://www.zomato.com/manila/red-buffalo-wing...   
6829  https://www.zomato.com/manila/greenwich-north-...   

                                         featured_image  has_online_delivery  \
0     https://b.zmtcdn.com/data/reviews_photos/971/2...                    0   
1     https://b.zmtcdn.com/data/res_imagery/18290970...                    0   
2     https://b.zmtcdn.com/data/pictures/9/6307689/b...                    0   
3     https://b.zmtcdn.com/data/reviews_photos/cc1/a...                    0   
4     https://b.zmtcdn.com/data/pictures/chains/8/63...                    0   
...                                                 ...                  ...   
6825                                                NaN                    0   
6826  http://b.zmtcdn.com/data/pictures/chains/0/630...                    0   
6827  https://b.zmtcdn.com/data/res_imagery/6300003_...                    0   
6828  https://b.zmtcdn.com/data/res_imagery/6309612_...                    0   
6829  https://b.zmtcdn.com/data/pictures/chains/2/63...                    0   

      has_table_booking  ...  country_id   latitude  \
0                     0  ...         162  14.560313   
1                     0  ...         162  14.552137   
2                     0  ...         162  14.563144   
3                     0  ...         162  14.633713   
4                     0  ...         162  14.546236   
...                 ...  ...         ...        ...   
6825                  0  ...         162  14.422261   
6826                  0  ...         162  14.756424   
6827                  0  ...         162  14.756406   
6828         

In [130]:
assert (6830, 35)==merged_df.shape

cols = ['id', 'address', 'city', 'city_id', 'country_id', 'latitude',
       'locality', 'locality_verbose', 'longitude', 'zipcode',
       'aggregate_rating', 'rating_color', 'rating_text', 'votes',
       'average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view',
       'is_delivering_now', 'is_table_reservation_supported',
       'is_zomato_book_res', 'menu_url', 'mezzo_provider', 'name',
       'photos_url', 'price_range', 'thumb', 'url']

assert sorted(merged_df.columns) == sorted(cols)

2. Create a function that returns the average aggregate rating for each city. Expected returned value is a Series datatype which contains name of city as index and average aggregate rating as value. **(2pts)**

In [131]:
def exercise_2(merged_df):
    return merged_df.groupby("city")["aggregate_rating"].agg("mean")

In [132]:
merged_df.groupby("city")["aggregate_rating"].agg("mean")

city
Caloocan City       1.953509
Las Piñas City      2.855738
Makati City         3.250512
Malabon City        1.558333
Mandaluyong City    2.951919
Manila              2.842652
Marikina City       2.672512
Muntinlupa City     2.962682
Parañaque City      3.130100
Pasay City          3.302075
Pasig City          2.916086
Pateros City        1.740000
Quezon City         2.887946
San Juan City       3.098462
Tagaytay City       2.959259
Taguig City         3.234105
Valenzuela City     1.890741
Name: aggregate_rating, dtype: float64

In [133]:
res_series = exercise_2(merged_df)
assert True == np.isclose(3.302075, res_series[res_series.index == 'Pasay City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.890741, res_series[res_series.index == 'Valenzuela City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.558333, res_series[res_series.index == 'Malabon City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]

3. Using the apply function of pandas convert the rating_text to a numeric equivalent. You are to introduce a new column in merged_df called `rating_numeric` which is based on the value from rating_text: **(3pts)**

Excellent = 5 <br>
Very Good = 4 <br>
Good = 3 <br>
Average = 2 <br>
Poor = 1 <br>
Not rated = 0 <br>

In [134]:
rating_dict = {"Excellent":5, "Very Good":4, "Good":3, "Average":2, "Poor":1, "Not rated":0}
merged_df["rating_numeric"] = merged_df["rating_text"].apply(lambda x: rating_dict.get(x))
merged_df[["name", "rating_numeric", "rating_text"]]

name  rating_numeric rating_text
0             Mendokoro Ramenba               5   Excellent
1            Soban K-Town Grill               5   Excellent
2               Filling Station               4   Very Good
3                Romantic Baboy               4   Very Good
4                NIU by Vikings               5   Excellent
...                         ...             ...         ...
6825              Trader Juan's               2     Average
6826                   Jollibee               2     Average
6827                 Goldilocks               0   Not rated
6828  Red Buffalo Wings & Pizza               3        Good
6829                  Greenwich               0   Not rated

[6830 rows x 3 columns]

In [135]:
assert 868 == merged_df[merged_df.rating_numeric == 0].shape[0]
assert True == ('rating_numeric' in merged_df.columns)

4. Create a function which filters the merged dataframe to only contain rows coming in from a specific city. Expected returned value is a dataframe **(2pts)**

In [136]:
merged_df
def exercise_4(merged_df, city):
    return merged_df[merged_df["city"] == city]

In [137]:
assert 'Makati City' == exercise_4(merged_df, 'Makati City').city.unique()[0]
assert (976, 36) == exercise_4(merged_df, 'Makati City').shape

assert 'Mandaluyong City' == exercise_4(merged_df, 'Mandaluyong City').city.unique()[0]
assert (443, 36) == exercise_4(merged_df, 'Mandaluyong City').shape

assert (0, 36) == exercise_4(merged_df, 'Quezon  City').shape

5. Create a function to filter the dataframe to only contain certain columns passed in the function. Expected returned value is a dataframe **(1pt)**

In [138]:
merged_df
def exercise_5(merged_df, cols_to_filter):
    return merged_df[cols_to_filter]

In [139]:
cols = ['name', 'cuisines', 'city', 'aggregate_rating']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 4) == exercise_5(merged_df, cols).shape

cols = ['average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 11) == exercise_5(merged_df, cols).shape


6.  What is the most common restaurant name in the dataset? **(2pts)** Write down your answer and show code used to come up with the answer.

In [140]:
merged_df.value_counts("name")

name
Starbucks                         99
Jollibee                          77
McDonald's                        66
KFC                               54
BonChon Chicken                   49
                                  ..
HotBox Kitchen                     1
Hotel Bar                          1
Hotties Gourmet                    1
House Manila - Remington Hotel     1
¡VENGO!                            1
Name: count, Length: 3286, dtype: int64

The most common restaurant name is Starbucks, with 99 occurrences.

7. Which restaurant has the most expensive average cost for two? **(2pts)** Write down your answer and show code used to come up with the answer.

In [141]:
merged_df[["name","average_cost_for_two"]].sort_values(by="average_cost_for_two", ascending=False)

name  average_cost_for_two
1713                           The Tasting Room                 10000
5      Spiral - Sofitel Philippine Plaza Manila                  6000
1728               Medley Buffet - Okada Manila                  5400
1750  Wolfgang's Steakhouse by Wolfgang Zwiener                  5100
2039  Wolfgang's Steakhouse by Wolfgang Zwiener                  5100
...                                         ...                   ...
6796                           Katherine's Cafe                     0
4714                            The Baking Crew                     0
6482                                   Toritake                     0
6192                                  Sari Salo                     0
6509                                   KapeTayo                     0

[6830 rows x 2 columns]

The restaurant with the most expensive average cost for two is The Tasting Room at 10,000 pesos.

8. How many restaurants do not have a zipcode? **(1pt)** Write down your answer and show code used to come up with the answer.

In [142]:
merged_df[merged_df["zipcode"].isna()]

id  average_cost_for_two book_url  \
0      6316125                  1000      NaN   
1     18409457                  1300      NaN   
3     18581637                  1000      NaN   
4      6315438                  3000      NaN   
5      6300010                  6000      NaN   
...        ...                   ...      ...   
6825  18635649                  1000      NaN   
6826   6318581                   400      NaN   
6827   6318578                   450      NaN   
6828   6309612                  1000      NaN   
6829   6318582                   600      NaN   

                                        cuisines currency  \
0                                Japanese, Ramen        P   
1                      Korean, Grill, Korean BBQ        P   
3                                         Korean        P   
4     Seafood, American, Mediterranean, Japanese        P   
5                        European, Asian, Indian        P   
...                                          ...      ...   
6825                              Deli, Filipino        P   
6826                                   Fast Food        P   
6827                            Bakery, Filipino        P   
6828                           American, Italian        P   
6829                            Pizza, Fast Food        P   

                          deeplink  \
0      zomato://restaurant/6316125   
1     zomato://restaurant/18409457   
3     zomato://restaurant/18581637   
4      zomato://restaurant/6315438   
5      zomato://restaurant/6300010   
...                            ...   
6825  zomato://restaurant/18635649   
6826   zomato://restaurant/6318581   
6827   zomato://restaurant/6318578   
6828   zomato://restaurant/6309612   
6829   zomato://restaurant/6318582   

                                             events_url  \
0     https://www.zomato.com/manila/mendokoro-ramenb...   
1     https://www.zomato.com/manila/soban-k-town-gri...   
3     https://www.zomato.com/manila/romantic-baboy-t...   
4     https://www.zomato.com/manila/niu-by-vikings-b...   
5     https://www.zomato.com/manila/spiral-sofitel-p...   
...                                                 ...   
6825  https://www.zomato.com/manila/trader-juans-fil...   
6826  https://www.zomato.com/manila/jollibee-north-c...   
6827  https://www.zomato.com/manila/goldilocks-north...   
6828  https://www.zomato.com/manila/red-buffalo-wing...   
6829  https://www.zomato.com/manila/greenwich-north-...   

                                         featured_image  has_online_delivery  \
0     https://b.zmtcdn.com/data/reviews_photos/971/2...                    0   
1     https://b.zmtcdn.com/data/res_imagery/18290970...                    0   
3     https://b.zmtcdn.com/data/reviews_photos/cc1/a...                    0   
4     https://b.zmtcdn.com/data/pictures/chains/8/63...                    0   
5     https://b.zmtcdn.com/data/res_imagery/6300010_...                    0   
...                                                 ...                  ...   
6825                                                NaN                    0   
6826  http://b.zmtcdn.com/data/pictures/chains/0/630...                    0   
6827  https://b.zmtcdn.com/data/res_imagery/6300003_...                    0   
6828  https://b.zmtcdn.com/data/res_imagery/6309612_...                    0   
6829  https://b.zmtcdn.com/data/pictures/chains/2/63...                    0   

      has_table_booking  ...   latitude  \
0                     0  ...  14.560313   
1                     0  ...  14.552137   
3                     0  ...  14.633713   
4                     0  ...  14.546236   
5                     0  ...  14.553256   
...                 ...  ...        ...   
6825                  0  ...  14.422261   
6826                  0  ...  14.756424   
6827                  0  ...  14.756406   
6828                  0  ...  14.610628   
6829                  0  ...  14.756382   

                                              

There are 6713 restaurants without a zipcode.

9. How many restaurants in the dataset allow online delivery? **(1pt)** Write down your answer and show code used to come up with the answer.

In [143]:
merged_df[merged_df["has_online_delivery"] == 1]

Empty DataFrame
Columns: [id, average_cost_for_two, book_url, cuisines, currency, deeplink, events_url, featured_image, has_online_delivery, has_table_booking, include_bogo_offers, is_book_form_web_view, is_delivering_now, is_table_reservation_supported, is_zomato_book_res, menu_url, mezzo_provider, name, photos_url, price_range, thumb, url, address, city, city_id, country_id, latitude, locality, locality_verbose, longitude, zipcode, aggregate_rating, rating_color, rating_text, votes, rating_numeric]
Index: []

[0 rows x 36 columns]

No restaurants support online delivery (sad!)

10. Which locality has more expensive food? Poblacion or Salcedo Village? **(3pts)** Write down your answer and show code used to come up with the answer.


In [144]:
pobla_salcedo_filter = merged_df[(merged_df["locality"] == "Poblacion") | (merged_df["locality"] == "Salcedo Village")]
pobla_salcedo_filter.groupby("locality")["average_cost_for_two"].agg("mean")

locality
Poblacion          898.214286
Salcedo Village    989.102564
Name: average_cost_for_two, dtype: float64

Salcedo Village has the more expensive food, where the aggregate average cost for two is 989.10 pesos.